### PR TITLE
Plugin/route53 AWS API MaxRetries parameter

### PR DESCRIPTION
**Pull request by rnsc**
### 1. Why is this pull request needed and what does it do?
This PR adds the possibility to tune the MaxRetries parameter of the AWS session created with the AWS SDK.

### 2. Which issues (if any) are related?
https://github.com/coredns/coredns/issues/4899

### 3. Which documentation changes (if any) need to be made?
I already updated the README for the plugin.

### 4. Does this introduce a backward incompatible change or deprecation?
No, it's a new parameter with a default.

### CODE REVIEW

File: plugin/route53/setup.go [hunk 1]
```diff
@@ -109,6 +115,20 @@ func setup(c *caddy.Controller) error {
 				} else {
 					return plugin.Error("route53", c.ArgErr())
 				}
+			case "aws_api_max_retries":
```
Comment (miekg) [inline plugin/route53/setup.go]: this may be better named 'max_retries'

File: plugin/route53/README.md [hunk 2]
```diff
@@ -22,42 +22,46 @@ route53 [ZONE:HOSTED_ZONE_ID...] {
...
-*   `refresh` can be used to control how long between record retrievals from Route 53. It requires
-    a duration string as a parameter to specify the duration between update cycles. Each update
-    cycle may result in many AWS API calls depending on how many domains use this plugin and how
-    many records are in each. Adjusting the update frequency may help reduce the potential of API
-    rate-limiting imposed by AWS.
+* `refresh` can be used to control how long between record retrievals from Route 53. It requires
+ a duration string as a parameter to specify the duration between update cycles. Each update
+ cycle may result in many AWS API calls depending on how many domains use this plugin and how
+ many records are in each. Adjusting the update frequency may help reduce the potential of API
+ rate-limiting imposed by AWS.
```
Comment (miekg) [inline plugin/route53/README.md]: what's up with all these unrelated changes? Please revert those

Comment (rnsc) [reply plugin/route53/README.md]: Right, my markdown linter in VSCode was complaining, thus me trying to fix it, I'll revert it, sorry

File: plugin/route53/setup.go (hunk 1 above)
Comment (rnsc) [reply plugin/route53/setup.go]: Done

File: plugin/route53/README.md (hunk 2 above)
Comment (rnsc) [reply plugin/route53/README.md]: Reverted!

**Comment by rnsc** [comment]
Not sure what happened with the reviewers addition following my conflict resolution... I'm sorry about the spam

**Comment by yongtang** [comment]
@rnsc Can you try rebase with the latest master (assuming `upstream` is pointing to coredns/coredns git):
```
git checkout plugin/route53-aws-api-retry-parameter
git fetch upstream master
git rebase upstream/master
git push -f origin plugin/route53-aws-api-retry-parameter
```

**Comment by rnsc** [comment]
> @rnsc Can you try rebase with the latest master (assuming `upstream` is pointing to coredns/coredns git):
> 
> ```
> git checkout plugin/route53-aws-api-retry-parameter
> git fetch upstream master
> git rebase upstream/master
> git push -f origin plugin/route53-aws-api-retry-parameter
> ```

Hello @yongtang I did just that, and I get conflicts on go.mod and go.sum, files which I haven't touched myself, wondering if I should fork from scratch?
image
image

**Comment by yongtang** [comment]
@rnsc You can try manually remove the duplicate locally to resolve the conflict. From the screenshot it seems to be purely duplicate so you can delete either one (after the build I guess go will pick up the latest one anyway).

**Comment by rnsc** [comment]
@yongtang Ok, I think I fixed it, thanks for the help!

File: plugin/route53/README.md [hunk 3]
```diff
@@ -49,6 +49,10 @@ route53 [ZONE:HOSTED_ZONE_ID...] {
     many records are in each. Adjusting the update frequency may help reduce the potential of API
     rate-limiting imposed by AWS.
 
+*   `max_retries` can be used to retry recoverable errors on the AWS API. By default errors
+    are not retried and thus, in case of throttling on the AWS API, the plugin doesn't update the zone.
+    The default value is 0, meaning that the pluging doesn't retry.
```
Comment (chrisohaver) [inline plugin/route53/README.md:54]:
```suggestion
    The default value is 0, meaning that the plugin doesn't retry.
```

Comment (chrisohaver) [reply plugin/route53/README.md:54]: Does this apply while establishing a connection, or while the connection is active,  or both?

What is considered a "recoverable" vs "unrecoverable" error?

Comment (rnsc) [reply plugin/route53/README.md:54]: @chrisohaver I believe this documentation describes this, maybe it should be mentioned here?The AWS Go documentation mentions "retryable errors", so my wording could be confusing because it doesn't match that:
https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/aws/retry

> By default the standard retryer uses the DefaultRetryables slice of IsErrorRetryable types to determine whether a given error is retryable. By default this list of retryables includes the following:
```raw
- Retrying errors that implement the RetryableError method, and return true.
- Connection Errors
  - Errors that implement a ConnectionError, Temporary, or Timeout method that return true.
  - Connection Reset Errors.
  - net.OpErr types that are dialing errors or are temporary.
  - HTTP Status Codes: 500, 502, 503, and 504.
- API Error Codes
  - RequestTimeout, RequestTimeoutException
  - Throttling, ThrottlingException, ThrottledException, RequestThrottledException, TooManyRequestsException,
    RequestThrottled, SlowDown, EC2ThrottledException
  - ProvisionedThroughputExceededException, RequestLimitExceeded, BandwidthLimitExceeded, LimitExceededException
  - TransactionInProgressException, PriorRequestNotComplete
```

Comment (rnsc) [reply plugin/route53/README.md:54]: And as far as I understood the documentation, the parameter is set when initialising the client connection

Comment (chrisohaver) [reply plugin/route53/README.md:54]: Thanks! ....  so I'm guessing it means pretty much any error other than an active connection refused or failed auth.  Maybe just don't mention "recoverable errors", since details of exactly what that means are a tad esoteric. e.g. 

> `max_retries` can be used to retry failed connections to the AWS API.

We should provide guidance here on when this should be used, and what value it should be set to.  Otherwise it may come across as a "magical number knob".

Are there drawbacks to setting this number very high? 

Is zero retries a sensible default?

Comment (chrisohaver) [reply plugin/route53/README.md:54]: IOW - does this need to be configurable?  Can we reasonably pick some magic number that would work.

Comment (rnsc) [reply plugin/route53/README.md:54]: @chrisohaver thank you for the better wording example, it does make sense to have it more generic.

In regards to defaults, as far as my testing went and how I understood the documentation, it's 0 by default already, thus my idea to define the variable and set it to 0 and have a non mandatory configuration option available to override it.

Now a disclaimer on the impact of setting that value should be written then, because due to the exponential backoff feature that is backed in the SDK, if you set the retries to, for example, 20, and you encounter a lot of errors, throttling, that could lead to a very long loading time.

Comment (rnsc) [reply plugin/route53/README.md:54]: Maybe adding a max value check in the code could be worth it, to avoid people trying huge values that would be detrimental in case of high errors rate on the API?

Comment (chrisohaver) [reply plugin/route53/README.md:54]: > if you set the retries to, for example, 20, and you encounter a lot of errors, throttling, that could lead to a very long loading time.

OK, but the alternative is CoreDNS exits ~immediately~ before starting.  

From a user standpoint, the actual number of retries is somewhat immaterial.  What actually matters is how long we would retry before giving up and aborting CoreDNS start up.

IMO, it might be better to avoid this by not requiring a successful API connection before CoreDNS starts.
A good behavior that doesn't require user config would be for CoreDNS to start, while the plugin retries connections in the background (logging errors on failure).

Comment (chrisohaver) [reply plugin/route53/README.md:54]: >> Is zero retries a sensible default?

> In regards to defaults, as far as my testing went and how I understood the documentation, it's 0 by default already, thus my idea to define the variable and set it to 0 and have a non mandatory configuration option available to override it.

I intended to question whether or not zero retries makes sense as the default behavior.  i.e. ...
Perhaps a value other than zero would make a better default.
And then it follows that if we can pick a default that would work well for everyone, we don't need to expose a knob.

Comment (rnsc) [reply plugin/route53/README.md:54]: @chrisohaver that's also fine with me, refactoring the plugin to support an async method to lazy load the configured zones is not really something I'd be comfortable with though (from a coding standpoint, I'd rather have someone with better Go knowledge take a look at it), I thought this was an ok middle ground solution before a better approach can be put in place, wdyt?

Comment (rnsc) [reply plugin/route53/README.md:54]: > > > Is zero retries a sensible default?
> 
> > In regards to defaults, as far as my testing went and how I understood the documentation, it's 0 by default already, thus my idea to define the variable and set it to 0 and have a non mandatory configuration option available to override it.
> 
> I intended to question whether or not zero retries makes sense as the default behavior. i.e. ... Perhaps a value other than zero would make a better default. And then it follows that if we can pick a default that would work well for everyone, we don't need to expose a knob.

Ah, right, I see what you mean! I'm also fine with a default low retry value without exposing the value to users, that's probably better!

**Comment by chrisohaver** [comment]
> A good behavior that doesn't require user config would be for CoreDNS to start, while the plugin retries connections in the background (logging errors on failure).

I think solving it this way would also fix #4894.

File: plugin/route53/README.md (hunk 3 above)
Comment (chrisohaver) [reply plugin/route53/README.md:54]: > I thought this was an ok middle ground solution before a better approach can be put in place, wdyt?

Yes, although ideally we shouldn't have to deprecate a user config when that happens, hence hoping we can pick a sane default without making it user configurable.

**Comment by chrisohaver** [comment]
I think we can just add a set number of retries, without adding a user config. e.g. 5 or something that would fix #4899, without hijacking start up too long if a connection to the API is never successful.
